### PR TITLE
createBall and createRobot from simBall and simRobot protobuf

### LINKED ISSUE #2328: Ball/RobotFilter: Translation functions from SimBall and SimRobot
### Description of the task
For the BallFilter and RobotFilter, we need to compare the ErForceSim SimBall and SimRobot(extlibs/er_force_sim/src/protobuf/world.proto) with our own Ball and Robot (software/world/[ball/robot].h) classes. The suggested approach is to write `Ball createBall (world::SimBall sim_ball)` and `Robot createRobot(world::SimRobot sim_robot)` functions in `proto/message_translation/er_force_world.[h/cpp]`. With the SimBall and SimRobot converted to our own classes, comparison can be easily done using the getters.

### Acceptance criteria

- [ ] createBall function
- [ ] createRobot function
- [ ] tests for new functions

### PR DESCRIPTION
**Pull request by jameschuang2002**
### Testing Done

Basic Conversion of SimBall and SimRobot

### Resolved Issues

resolves #2328 
### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [x] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [x] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

File: src/proto/message_translation/er_force_world.h [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+#include "software/world/ball.h"
+#include "software/world/ball_state.h"
+#include "software/world/robot.h"
+#include "software/world/robot_state.h"
+
+/**
+ * convert a er force simulator ball to a ball object
+ *
+ * @param sim_ball: the ball information retrieved from the er force simulator
+ *
```
Comment (jonathanlew) [inline src/proto/message_translation/er_force_world.h]: missing param for timestamp

File: src/proto/message_translation/er_force_world.h [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+ * @return the corresponding ball object to the er simulator ball
+ */
+
+Ball createBall(world::SimBall sim_ball, Timestamp timestamp);
+
+/**
+ * Convert a er force simulator robot to a robot object
+ *
+ * @param sim_robot: the robot information retrieved from the er force simulator
```
Comment (jonathanlew) [inline src/proto/message_translation/er_force_world.h:21]: missing param for timestamp

File: src/proto/message_translation/er_force_world.cpp [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+    const BallState state(position, velocity);
+    const Ball ball(state, timestamp);
+    return ball;
+}
+Robot createRobot(world::SimRobot sim_robot, Timestamp timestamp)
+{
+    const RobotId id(sim_robot.id());
+    const Point position(sim_robot.p_x(), sim_robot.p_y());
+    const Vector velocity(sim_robot.v_x(), sim_robot.v_y());
+    double angle = 2 * Angle::acos(sim_robot.rotation().real()).toRadians();
```
Comment (jonathanlew) [inline src/proto/message_translation/er_force_world.cpp]: Shouldn't this involve the non real parts of the quaternion as well? https://en.wikipedia.org/wiki/Quaternions_and_spatial_rotation

**Review by jonathanlew — CHANGES_REQUESTED** [review]
Looks good, just a few comments. Also please rename the title of this PR

File: src/proto/message_translation/er_force_world.cpp [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+    const Point position(sim_robot.p_x(), sim_robot.p_y());
+    const Vector velocity(sim_robot.v_x(), sim_robot.v_y());
+
+    /* adopted conversion from
+     * https://en.wikipedia.org/wiki/Conversion_between_quaternions_and_Euler_angles */
+    world::Quaternion q = sim_robot.rotation();
+    double siny_cosp    = 2 * (q.real() * q.k() + q.i() * q.j());
+    double cosy_cosp    = 1 - 2 * (q.j() * q.j() + q.k() * q.k());
+    double angle        = Angle::atan(siny_cosp / cosy_cosp).toRadians();
+    const RobotState state(position, velocity, Angle::fromRadians(angle),
```
Comment (jonathanlew) [inline src/proto/message_translation/er_force_world.cpp]:
```suggestion
    Angle angle        = Angle::atan(siny_cosp / cosy_cosp);
    const RobotState state(position, velocity, angle,
```

File: src/proto/message_translation/er_force_world_test.cpp [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+    sim_robot->set_v_y(10.0);
+    sim_robot->set_v_z(12.0);
+    sim_robot->set_r_x(1.0);
+    sim_robot->set_r_y(1.0);
+    sim_robot->set_r_z(5.0);
+
+    const Robot test_robot = createRobot(*sim_robot, Timestamp::fromSeconds(0));
+    const Point expected_pos(sim_robot->p_x(), sim_robot->p_y());
+    const Vector expected_vel(sim_robot->v_x(), sim_robot->v_y());
+    RobotState expected_state(expected_pos, expected_vel, Angle::fromRadians(0),
```
Comment (jonathanlew) [inline src/proto/message_translation/er_force_world_test.cpp]: can we have a test where the expected angle is nonzero please?

File: src/proto/message_translation/er_force_world_test.cpp [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+                              Angle::fromDegrees(45));
+    EXPECT_EQ(0, test_robot.id());
+    EXPECT_TRUE(TestUtil::equalWithinTolerance(test_robot.currentState(), expected_state,
+                                               1e-6, Angle::fromDegrees(5)));
```
Comment (jonathanlew) [inline src/proto/message_translation/er_force_world_test.cpp]: I think the degrees tolerance can be smaller like 1 degree or less

**Review by jonathanlew — APPROVED** [review]
LGTM assuming the quaternion conversion is correct

**Review by akhilveeraghanta — APPROVED** [review]
lgtm!

**Pull request merged**